pkg/resources/validation/namespace/namespace.go

```diff
@@ -0,0 +1,106 @@
+	admissionv1 "k8s.io/api/admission/v1"
+	admissionregistrationv1 "k8s.io/api/admissionregistration/v1"
+	v1 "k8s.io/api/authorization/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	authorizationv1 "k8s.io/client-go/kubernetes/typed/authorization/v1"
+	"k8s.io/utils/trace"
+)
+
+var gvr = schema.GroupVersionResource{
+	Version:  "v1",
+	Resource: "namespaces",
+}
+
+// Validator validates the namespace admission request
+type Validator struct {
+	sar authorizationv1.SubjectAccessReviewInterface
+}
+
+// NewValidator returns a new validator used for validation of namespace requests.
+func NewValidator(sar authorizationv1.SubjectAccessReviewInterface) *Validator {
+	return &Validator{
+		sar: sar,
+	}
+}
+
+// GVR returns the GroupVersionKind for this CRD.
+func (v *Validator) GVR() schema.GroupVersionResource {
+	return gvr
+}
+
+// Operations returns list of operations handled by this validator.
+func (v *Validator) Operations() []admissionregistrationv1.OperationType {
+	return []admissionregistrationv1.OperationType{admissionregistrationv1.Update}
+}
+
+// ValidatingWebhook returns the ValidatingWebhook used for this CRD.
+func (v *Validator) ValidatingWebhook(clientConfig admissionregistrationv1.WebhookClientConfig) *admissionregistrationv1.ValidatingWebhook {
+	// Note that namespaces are actually CLUSTER scoped
+	return admission.NewDefaultValidationWebhook(v, clientConfig, admissionregistrationv1.ClusterScope)
+}
+
+// Admit is the entrypoint for the validator.
+// Admit will return an error if it unable to process the request.
+func (v *Validator) Admit(request *admission.Request) (*admissionv1.AdmissionResponse, error) {
+	listTrace := trace.New("Namespace Admit", trace.Field{Key: "user", Value: request.UserInfo.Username})
+	defer listTrace.LogIfLong(2 * time.Second)
+
+	response := &admissionv1.AdmissionResponse{}
+
+	oldns, ns, err := objectsv1.NamespaceOldAndNewFromRequest(&request.AdmissionRequest)
+	if err != nil {
+		response.Result.Code = http.StatusBadRequest
+		return response, fmt.Errorf("failed to decode namespace from request: %w", err)
+	}
+
+	if oldns == nil {
+		return nil, nil
+	}
+	// Is the request attempting to modify the special PSA labels (enforce, warn, audit)
+	// If it isn't, we're done
+	// If it is, we then need to check to see if they should be allowed
+	if !validation.IsUpdatingPSAConfig(oldns.Labels, ns.Labels) {
+		response.Allowed = true
+		return response, nil
+	}
+
+	resp, err := v.sar.Create(request.Context, &v1.SubjectAccessReview{
+		Spec: v1.SubjectAccessReviewSpec{
+			ResourceAttributes: &v1.ResourceAttributes{
+				Verb:     "updatepsa",
+				Group:    "management.cattle.io",
+				Version:  "v3",
+				Resource: "projects",
+			},
+			User:   request.UserInfo.Username,
+			Groups: request.UserInfo.Groups,
+			UID:    request.UserInfo.UID,
+		},
+	}, metav1.CreateOptions{})
+	if err != nil {
+		return response, err
```

Comment:
1. Wrap the SAR error.


2. Nit: I like to explicitly return nil so that the caller knows the response is not valid. Currently, it is not used by the caller if the error is not nil, so I made this a nit. But I think it keeps the code easier to read and consistent.
```suggestion
		return nil, err
```